### PR TITLE
(PUP-11931) Resolve Lint rubocops

### PR DESCRIPTION
**Pull request by joshcooper**
The commits marked (bug) are actual bugs that should be backported to 7.x. Most of these are mechanical or easy to resolve. Here are some notable cops resolved in this PR:

### Lint/ReturnInVoidContext & Lint/Void
By default, `initialize` doesn't return a value unless `new` is also overridden. Also setter methods always return the argument passed to the setter, not what the setter returns.

### Lint/UriRegexp
`URI.regexp` is obsolete, use `URI::DEFAULT_PARSER.make_regexp` instead.

### Lint/MixedRegexpCaptureTypes
If a regexp contains both named and numbered capture groups, then the numbered capture groups are ignored. We weren't relying on that behavior, but it could be surprising. Explicitly use `?:` to indlcate the group is non-capturing.

### Lint/ShadowedArgument
Puppet was using a complicated `Array#reduce` construct to find the first matching element

### Lint/StructNewOverride
`Struct.new(:foo)` defines a new class with accessors for `foo`. Our server-side request struct is accidentally overriding the builtin `Object#method`

The following cops are still todo because they're not easy to resolve. I'll file a separate ticket about them:

```
Lint/ConstantDefinitionInBlock
Lint/MissingSuper
Lint/NestedMethodDefinition
Lint/RescueException
Lint/SuppressedException
Lint/ToJSON
Lint/UnusedMethodArgument
```

### CODE REVIEW

File: lib/puppet/module_tool/metadata.rb [hunk 1]
```diff
@@ -97,7 +97,7 @@ def to_hash
     end
     alias :to_data_hash :to_hash
 
-    def to_json
+    def to_json(*_args)
```
Comment (joshcooper) [inline lib/puppet/module_tool/metadata.rb:100]: Revisit this

File: lib/puppet/util.rb [hunk 2]
```diff
@@ -332,7 +332,7 @@ def uri_to_path(uri)
   end
   module_function :uri_to_path
 
-  RFC_3986_URI_REGEX = /^(?<scheme>([^:\/?#]+):)?(?<authority>\/\/([^\/?#]*))?(?<path>[^?#]*)(\?(?<query>[^#]*))?(#(?<fragment>.*))?$/
+  RFC_3986_URI_REGEX = /^(?<scheme>(?:[^:\/?#]+):)?(?<authority>\/\/(?:[^\/?#]*))?(?<path>[^?#]*)(?:\?(?<query>[^#]*))?(?:#(?<fragment>.*))?$/
```
Comment (joshcooper) [inline lib/puppet/util.rb:335]: mention this in PR description

File: lib/puppet/util/profiler/aggregate.rb [hunk 3]
```diff
@@ -16,10 +16,6 @@ def shutdown()
     @logger.call("----------------------------")
   end
 
-  def do_start(description, metric_id)
-    super(description, metric_id)
-  end
-
```
Comment (joshcooper) [inline lib/puppet/util/profiler/aggregate.rb:22]: add comment instead of removing

File: lib/puppet/module_tool/metadata.rb (hunk 1 above)
Comment (joshcooper) [reply lib/puppet/module_tool/metadata.rb]: I removed this commit, it's definitely a bug, but not sure how to resolve it right now.

**Review by AriaXLi — APPROVED** [review]

**Pull request merged**